starter-kits/credential-registry/server/tob-api/icat_hooks/serializers/hooks.py

```diff
@@ -143,14 +156,78 @@ def validate_credential_type(self, value):
+        # validate subscription parameters
+        subscription_type = (
+            data["subscription_type"] if "subscription_type" in data else None
+        )
+        topic_source_id = data["topic_source_id"] if "topic_source_id" in data else None
+        credential_type = data["credential_type"] if "credential_type" in data else None
+        if subscription_type == "New" and topic_source_id is None:
+            raise serializers.ValidationError(
+                "A topic id is required for subscription of type 'New'"
+            )
+        if subscription_type == "Stream" and (
+            topic_source_id is None or credential_type is None
+        ):
+            raise serializers.ValidationError(
+                "A topic id and a credential type are required for subscription of type 'Topic'"
+            )
+
+        # get current user from url
+        uri = self.context.get("request").build_absolute_uri()
+        context_user = uri.split("registration/")[1].split("/")[0]
+        v2_user = User.objects.get(username=context_user)
+
+        # validate hook url and token
+        hook_url = data["target_url"] if "target_url" in data else None
+        hook_token = data["hook_token"] if "hook_token" in data else None
+        hook_user = HookUser.objects.get(user_id=v2_user.id)
+
+        if hook_url is None:
+            if hook_user.target_url is None:
+                raise serializers.ValidationError(
+                    "A target_url must be specified, no default value set in registration."
+                )
+            else:
+                data['target_url'] = hook_user.target_url
+
+        if hook_token is None:
+            if hook_user.hook_token is None:
+                raise serializers.ValidationError(
+                    "A hook-token must be specified, no default value set in registration."
+                )
+            else:
+                data['hook_token'] = hook_user.hook_token
+
+        # check hook url is valid
+        self.check_live_url(data['target_url'], data['hook_token'])
+
+        return data
+
+    def check_live_url(self, target_url, hook_token):
+        if target_url is not None and hook_token is not None:
+            head = {"Authorization": "Bearer " + hook_token}
+            data = {"subscription": "test"}
+            response = requests.post(target_url, json=data, headers=head)
+
+            if response.status_code != requests.codes.ok:
+                raise serializers.ValidationError(
+                    "The url {} does not appear to be valid.".format(target_url)
+                )
+
+        return "SUCCESS"
+
     def create(self, validated_data):
         """
         Create and return a new instance, given the validated data.
         """
         # note owner is assigned in the view
-        credential_type = CredentialType.objects.filter(
-            schema__name=validated_data["credential_type"]
-        ).first()
+        credential_type = None
```

Comment:
This is necessary so that _credential_type_ won't be required for subscriptions of type "New".